### DISCUSSION TITLE
Webhook Retries

### DISCUSSION THREAD

**Discussion by Wamy-Dev** [supabase/supabase#17664] [Feature Requests]
Hey there! I am wanting a new feature that doesn't seem to be present in supabase regarding webhooks. Sometimes my API can get backed up or is down for a few seconds, and webhooks may fail. Some functions rely on webhooks to work and provide accurate data to customers, so like Stripe does, I would like an automated retry system. 

If a webhook fails, it will automatically try in a time like 30 minutes. If that fails, it retries in an hour, maybe to 5 retries total. That way, no webhook is missed.

**Comment by brightertools** [comment]
I would also like to know if retries are done, when we implement webhooks, we run the retries on an exponential scale that tries with increasing periods, that ensure resilience to up to 24 hours, in case of connectivity downtime due to updates, issues etc. I would expect some sort of retry mechanism.

**Reply by la289** [reply]
Same - Im nervous to try this, but I found this documentation which describes one way to build the retries in: https://github.com/supabase/pg_net?tab=readme-ov-file#retrying-failed-requests

Seems like something that would be useful out the box with supabase

**Reply by ric2z** [reply]
> Same - Im nervous to try this, but I found this documentation which describes one way to build the retries in: https://github.com/supabase/pg_net?tab=readme-ov-file#retrying-failed-requests
> 
> Seems like something that would be useful out the box with supabase

Trying to implement this, we have the wrapper and the function. But it's not clear to me when if we have to create another trigger to make it work. Any help?

**Comment by Rubenburdin** [comment]
Hey ric2z, coming back to this in 2025, the pg_net retry pattern still requires you to build the orchestration yourself, which is the frustrating part.

To answer the question directly: yes, you need a separate mechanism to trigger the retry. The wrapper function from the docs handles the logic of "check response, retry if failed," but something has to call that wrapper on a schedule. Most folks use pg_cron for this. You'd create a cron job that periodically scans net._http_response for failed requests (status >= 400 or null) and re-fires them.

A minimal setup looks something like this: create a function that queries net._http_response for recent failures, extracts the original request details, and calls net.http_post again with the same payload. Then schedule it with SELECT cron.schedule('retry-failed-webhooks', '*/5 * * * *', 'SELECT retry_failed_requests();'); to run every 5 minutes. You'll want to track retry count somewheere (a simple counter column or separate table) so you don't retry forever.

The pg_cron extension should already be available on most Supabase projects these days. You can check with SELECT * FROM pg_extension WHERE extname = 'pg_cron'; (https://supabase.com/docs/guides/database/extensions/pg_cron). (https://supabase.com/docs/guides/cron)
It's more wiring than it should be for something this common, but once set up it's pretty reliable.
